RandomStateAntipattern.md

```diff
@@ -0,0 +1,383 @@
+manage it.  But our tasks continue to get harder by insisting on referential transparency.
+
+Now suppose we want to keep building up, creating these:
+
+```
+case class Bar(id: Long, b: Byte, z: Boolean) {}
+
+case class Baz(id: Long, foo: Foo, bar: Bar) {}
+```
+
+But now suppose we want to use yet another source of random numbers that is
+the only one that can produce `Byte`s, and which we also will use to make booleans.
+
+```
+trait Rn3 extends Rng {
+  def nextByte = (nextLong >>> 56).toByte
+}
+
+val rrr = new Rng(37) with Rn3 {}
+```
+
+No problem.
+
+And now we just create the things we want taking the random number streams we ought to:
+
+```
+def mkBarMut(rId: Rng with IdR, rB: Rn3) =
+  Bar(rId.nextLong, rB.nextByte, rB.nextBool)
+
+def mkBazMut(rId: Rng with IdR, rTf: Rng with TfR, rB: Rn3) =
+  Baz(rId.nextLong, mkFooMut(rId, rTf), mkBarMut(rId, rB))
+```
+
+There's one short obvious way to do it, and any simple error is a compile error.  (With the proviso, again,
+that we haven't made the streams only able to produce the data type we want.)
+
+I'm going to post the `State` version without describing it blow-by-blow.
+
+```
+type Rng3State[A] = State[(Long, Long, Long), A]
+
+def nextFromOneOf3[A](r: RngState[A]): Rng3State[A] = State{ s =>
+  val (sid2, a) = r.run(s._1).value
+  (s.copy(_1 = sid2), a)
+}
+
+def nextFromFooish[A](r2: Rng2State[A]): Rng3State[A] = State{ s =>
+  val (nx, a) = r2.run((s._1, s._2)).value
+  (s.copy(_1 = nx._1, _2 = nx._2), a)
+}
+
+def nextFromBarish[A](r2: Rng2State[A]): Rng3State[A] = State{ s =>
+  val (nx, a) = r2.run((s._1, s._3)).value
+  (s.copy(_1 = nx._1, _3 = nx._2), a)
+}
+
+// We're not even enforcing stream-type-safety here!
+val nextByte = nextLong.map(l => (l >>> 56).toByte)
+
+val mkBarFp =
+  for {
+    id <- nextFromOne(nextLong)
+    b <- nextFromTwo(nextByte)
+    z <- nextFromTwo(nextBool)
+  } yield Bar(id, b, z)
+
+val mkBazFp =
+  for {
+    id <- nextFromOneOf3(nextLong)
+    foo <- nextFromFooish(mkFooFp)
+    bar <- nextFromBarish(mkBarFp)
+  } yield Baz(id, foo, bar)
+```
+
+Whew!  It works, but to keep it even this short we've given up some parts of the
+type-safety of the streams (everyone can create `Byte` now), we've got growing
+amounts of fiddly state packing and unpacking, and we're increasingly repeating
+ourselves in `nextWhatever(makeWhatever)` pairs.
+
+And what do we gain for all this?  Only that we allow ourselves to forget the
+distinction between producing and storing a value, and that `Rng` does the former.
+
+## Debriefing
+
+`State` is a clumsy, unscalable mechanism for carrying state along with a
+computation, but it does royally well at actually enforcing that the state is
+carried along.
+
+In the case of random numbers, the problem is solved _autonomously_ by a mutable
+class that takes care of the state on its own.  Not all state problems are like
+this.  But many, including random number generation, are.  This separability,
```

Comment:
Maybe I should give more examples.  Counters, timestamps, and lazily loaded or computed data all have this property.  Iterators of various sorts are more finicky and generally don't have this property.  State machines are a separate case that I'll deal with in as another antipattern (in brief: they don't really have this property, but by taking on the burden of managing difficult state transitions autonomously, you can greatly simplify your interactions with them).

Reply:
I'm already less sure for counters, but random numbers and timestamps is definitely not something that' s done with `State`. 
The name `State` is not the best, because it gives the impression that it's the solution for dealing with state functionally. But what side-effectful programming refers to as "state" actually encompasses several different patterns, and `State` is only a solution to some of them (and a pretty good one if you incorporate the lens-y thing for modularity).

You can see that in my `PureRng` example. That being said, the title of the article is actually correct, if you interpret it as "that's not how you do this in FP", rather than "that's how you do this in FP, and it's not good" ;)